Sources/TSCBasic/Downloader.swift

```diff
@@ -0,0 +1,136 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+#if canImport(FoundationNetworking)
+import FoundationNetworking
+#endif
+
+/// Error enumeration that can be returned by a `Downloader` conforming type.
+public enum DownloaderError: Error {
+
+    /// Error thrown when the downloader fails to establish a connection to the server.
+    case clientError(error: Error)
+
+    /// Error thrown when the downloader received an invalid status code from the server.
+    case serverError(statusCode: Int)
+}
+
+/// The `Downloader` protocol abstract away the download of a file with a progress report.
+public protocol Downloader {
```

Comment:
I think we should put this in TSCUtility.

Reply:
Can you explain why? I think I don’t understand the difference between the two then.

Reply:
There isn't a well-defined boundary but this feel like a higher-level utility that probably won't be relevant to every project that uses TSC. For example, llbuild or SourceKit-LSP will most likely not need it.

Reply:
Moved to TSCUtility